core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -364,18 +362,37 @@ boolean forceRefresh() {
             executor,
             metadataFuture,
             sslContextFuture);
-    refreshFuture.addListener(
-        () -> {
-          synchronized (instanceDataGuard) {
-            // update currentInstanceData with the most recent results
-            currentInstanceData = refreshFuture;
-            // schedule a replacement before the SSLContext expires;
-            nextInstanceData = executor
-                .schedule(this::performRefresh, secondsUntilRefresh(),
-                    TimeUnit.SECONDS);
+    Futures.addCallback(refreshFuture,
+        new FutureCallback<InstanceData>() {
+          public void onSuccess(InstanceData instanceData) {
+            synchronized (instanceDataGuard) {
+              // update currentInstanceData with the most recent results
+              currentInstanceData = refreshFuture;
+              // schedule a replacement before the SSLContext expires;
+              nextInstanceData = executor
+                  .schedule(() -> performRefresh(),
+                      secondsUntilRefresh(),
+                      TimeUnit.SECONDS);
+            }
           }
-        },
-        executor);
+
+          public void onFailure(Throwable t) {
+            logger.log(Level.WARNING,
+                "An error occurred while performing refresh. Retrying immediately.", t);
+            if (getInstanceData().getExpiration().toInstant().isBefore(Instant.now())) {
```

Comment:
We need to be careful here. There are at least two problems I see with the current code.
1. If "current" is in a bad state (e.i. error) it'll throw an (maybe unrelated) error and nothing past this will happen
2. if "current" _is_ this thread (because we've already forced refresh), this will - well I'm not sure. Block forever?

Reply:
I don't think we'd ever get into a position where current is still pending, since this callback happens after refreshFuture has completed. The first problem is something to think about though. how can we check the validity of the instance data here, without potentially throwing errors. Would a try/catch work, where we try `getInstanceData` and if we get an error, we replace current

Reply:
Yea, I think I would go with a try/catch around it. Just to clarify - you are sure that the callback will trigger after the future is completed?

Reply:
One thing to think about - I think we actually want the replace of `current` and `next` to happen before the future is returned. Otherwise you may introduce a race condition where we return, fail to connect, and force refresh before the callback is executed

Reply:
Yeah, the callback will definitely trigger after the future is completed. I get what you're saying with the race condition, where a force refresh could be triggered before the callback, but how can we replace `current` before we know the future's result if our goal is to avoid replacing a valid `InstanceData` with an invalid one?

I think in the case where a force refresh is triggered before the callback, the worst that would happen is that we get an extra refresh attempt a minute later, since the rate limiter will delay the force refresh anyway. If the refresh attempt is successful, then we'd end up replacing a valid result with another (hopefully) valid result, or wait for another refresh attempt to fix it.

Reply:
To clarify, I think it's preferable if whatever the callback behavior is (replacing current or just rescheduling) before the future counts as "done" .

Reply:
Can you use `[thenApply`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#thenApply-java.util.function.Function-) for that?